Alerting: New notification policies view

gillesdemey opened the pull request:
**What is this feature?**

This PR introduces a new notification policies view; most notable changes:

* The new design shows the entire tree, no longer required to keep checking policies for child policies
* Show the alert instances currently handled by a particular route
* Visual aid to show if `continue matching siblings` is configured for a route
* Meta data is added to the overview, no longer needed to uncollapse a policy
* Shows what properties are inherited from their parent policy
* Shows warnings for policies that match all and have no `continue` flag set (these might be swallowing alerts)
* Shows if there are any contact point errors for the route
* Colors are the same for matchers with the same key, making it easier to see what routes will be matching on the same label key
* Adds a tab navigation for policies and mute timings
* Highlights routes in the tree when filtering for a specific contact point (this is now a `<Select>` component)

Hover states (not shown) to show more info about

* What integrations are configured for a contact point
* What properties are inherited and their values

<details>
<summary><strong>Before</strong></summary>
<img width="1024" alt="image" src="https://user-images.githubusercontent.com/868844/214123985-c33b49ef-956a-4486-8d7d-680c0833da5d.png">
</details>

<details open>
<summary><strong>After</strong></summary>
<img width="1163" alt="image" src="https://user-images.githubusercontent.com/868844/221187715-4275c2a0-8f61-49d9-a94d-0c9a89057c47.png">
</details>

**Why do we need this feature?**

The previous notification policies view was not showing a holistic view of the routing and required an operator to piece things together manually by inspecting each individual policy.

**Special notes for your reviewer**:

This PR is pretty large – I propose going through this in a separate mob session or taking your time over a few smaller review sessions :)

I've also moved (and renamed) several files from `amroutes` to `notification-policies` to help keep the code structure in line with the UI.

Setting this to draft while I update the tests that I broke and figure out some small UI issues.

Fixes

https://github.com/grafana/grafana/issues/62650
https://github.com/grafana/grafana/issues/39143
https://github.com/grafana/grafana/issues/46919

gillesdemey committed:
add route matching algorithm utility

gillesdemey committed:
WIP matching alert groups

still a bug with child policies

gillesdemey committed:
fix incorrect child policies match

gillesdemey committed:
fix route matching

pass currentRoute so we don't lose track of the tree cursor

gillesdemey committed:
rename instances to instance groups

gillesdemey committed:
wip UI

gillesdemey committed:
Add buttons and tab layout

gillesdemey committed:
hook up to actual AM config

gillesdemey committed:
normalize object matchers

gillesdemey committed:
hook up modals

gillesdemey committed:
fix continue matching

gillesdemey committed:
hook up edit / add form

gillesdemey committed:
small refactoring

gillesdemey committed:
hook up form inputs

gillesdemey committed:
fix contact point link

gillesdemey committed:
fix loading and error states

gillesdemey committed:
make editing work

gillesdemey committed:
small fixes

gillesdemey committed:
feat: implement contact point hover

gillesdemey committed:
use root form for default policy

gillesdemey committed:
show multiple warnings

gillesdemey committed:
feat: implement delete route

gillesdemey committed:
small edits

gillesdemey committed:
feat: implement "add policy"

gillesdemey committed:
tweaks

gillesdemey committed:
fix editing grafana managed routes

gillesdemey committed:
fix updating policies

gillesdemey committed:
minor form tweaks

gillesdemey committed:
add loading modals

gillesdemey committed:
make hover card arrow work for loght theme

gillesdemey committed:
show full inherited properties list for each node

gillesdemey committed:
feat: implement contact point errors

gillesdemey committed:
small errors fix

gillesdemey committed:
refactor getContactPointErrors

gillesdemey committed:
refactor: split up component

gillesdemey committed:
Merge branch 'main' into alerting/notification-policies-next

gillesdemey committed:
WIP bring back alert instances

gillesdemey committed:
fix modal close issue

gillesdemey committed:
pair programming and bug squashing 💪

gillesdemey committed:
use a dropdown buttongroup for actions

gillesdemey committed:
chore: update text

gillesdemey committed:
adds hovercard for additional matchers

gillesdemey committed:
adds default timing options

gillesdemey committed:
add simple contact point filtering algorithm

gillesdemey committed:
feat: adds summary for matching instances

gillesdemey committed:
cleanup

gillesdemey committed:
add navigation query param for mute timings

gillesdemey committed:
allow unsetting the notification policy

gillesdemey committed:
Convert mute timing hover card to links

gillesdemey committed:
update icon bundle

gillesdemey committed:
update copy

gillesdemey committed:
feat: implement label filtering

gillesdemey committed:
file cleanup

gillesdemey committed:
simplify component, add test-ids

gillesdemey committed:
adds test for policies

gillesdemey committed:
improve error display for multiple contact point types

gillesdemey committed:
more tests

gillesdemey committed:
simplify the route props

gillesdemey committed:
adds test for normalizing matchers

gillesdemey committed:
Merge branch 'main' into alerting/notification-policies-next

gillesdemey committed:
skipping broken tests for now

I'll enable these again when I find a better way to test notification policies e2e

gillesdemey committed:
Merge branch 'main' into alerting/notification-policies-next

gillesdemey committed:
update delivery warning component

gillesdemey committed:
reset file

gillesdemey committed:
remove icons from bundle

gillesdemey committed:
reset popover controller changes

gotjosh commented:
😱 ❤️

konrad147 commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -0,0 +1,660 @@
+  const numberOfAlertInstances = sumBy(matchingAlertGroups, (group) => group.alerts.length);
+
+  // TODO dead branch detection, warnings for all sort of configs that won't work or will never be activated
+  return (
```
Should we add more tailor-made components into this one?
The number of conditions and branches make the logic quite difficult to follow so maybe limiting the amount of UI code would help
e.g. Instead of 
```
{errors.length > 0 && (
                <HoverCard
                  arrow
                  placement="top"
                  content={
                    <Stack direction="column" gap={0.5}>
                      {errors.map((error) => (
                        <Fragment key={uniqueId()}>{error}</Fragment>
                      ))}
                    </Stack>
                  }
                >
                  <span>
                    <Badge icon="exclamation-circle" color="red" text={pluralize('error', errors.length, true)} />
                  </span>
                </HoverCard>
              )}
```
have sth like
```
{errors.length > 0 && <PolicyErrors errors={errors> />}

gillesdemey replied:
Yea I totally agree, I haven't spent enough time figure out what abstractions we can make here – will try to clean this up a bit.

gillesdemey replied:
I've cleaned up the policy component – right now the sub components still live in the same file (since it's very likely that if we change the policy component those will have to be updated too, I prefer to keep code that changes often together in the same file.

konrad147 commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -0,0 +1,660 @@
+    errors.push(error);
+  });
+
+  const childPolicies = currentRoute.routes ?? [];
```
How about extracting some part of the logic from the component into a function when we can test it in isolation?

konrad147 commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -0,0 +1,660 @@
+
+const getStyles = (theme: GrafanaTheme2) => ({
+  matcher: (label: string) => {
+    const { color, borderColor } = getTagColorsFromName(label);
```
Won't css performance suffer from this approach?

gillesdemey replied:
Yea it might, I believe this is negligible though and I'm not quite sure how to improve this – will extracting this in to its own function impact the performance of rendering the other styles / classNames?

konrad147 commented on public/app/features/alerting/unified/state/actions.ts:
```diff
@@ -547,6 +547,7 @@ export const updateAlertManagerConfigAction = createAsyncThunk<void, UpdateAlert
     withAppEvents(
       withSerializedError(
         (async () => {
+          // TODO there must be a better way here than to dispatch another fetch as this causes re-rendering :(
```
Not for this PR but AlertmanagerConfigAction is pretty straightforward so this is a good candidate to be refactored into RTK Query

konrad147 commented on public/app/features/alerting/unified/utils/misc.ts:
```diff
@@ -35,6 +35,16 @@ export function createExploreLink(dataSourceName: string, query: string) {
 }
 
+export function createContactPointLink(contactPoint: string, alertManagerSourceName = ''): string {
+  return `/alerting/notifications/receivers/${encodeURIComponent(contactPoint)}/edit?alertmanager=${encodeURIComponent(
```
Please use `createUrl` helper that takes care of the case when Grafana is hosted on a subpath

konrad147 commented on public/app/features/alerting/unified/utils/misc.ts:
```diff
@@ -35,6 +35,16 @@ export function createExploreLink(dataSourceName: string, query: string) {
+}
+
+export function createMuteTimingLink(muteTimingName: string, alertManagerSourceName = ''): string {
+  return `/alerting/routes/mute-timing/edit?muteName=${muteTimingName}&alertmanager=${alertManagerSourceName}`;
```
Same as above

konrad147 reviewed (commented):
A first round of review 😄

VikaCep commented on public/app/features/alerting/unified/components/notification-policies/Filters.tsx:
```diff
@@ -0,0 +1,172 @@
+          width={46}
+        />
+      </Field>
+      <Field label="Search by contact point" style={{ marginBottom: 0 }}>
```
Shouldn't it have its own class instead of using an inline style?

gillesdemey replied:
Yea it should, will update!

VikaCep commented on public/app/features/alerting/unified/components/notification-policies/Policy.test.tsx:
```diff
@@ -0,0 +1,257 @@
+    await userEvent.click(within(defaultPolicy).getByTestId('more-actions'));
+    expect(await screen.queryByRole('menuitem', { name: 'Delete' })).not.toBeInTheDocument();
+
+    // default policy should show the metadata
```
either missing assertion or extra comment here 😄

gillesdemey replied:
This comment is extra, it basically says "we're going to make some assertions on metadata of the default policy from here on" but I get why that would be confusing since the next line is another comment about what we're checking specifically 😅

VikaCep commented on public/app/features/alerting/unified/utils/amroutes.ts:
```diff
@@ -87,72 +88,118 @@ export const emptyRoute: FormAmRoute = {
+  return matchers;
+};
+
+// add unique identifiers to each route in the route tree, that way we can figure out what route we've edited / deleted
```
useful comment! 👏

VikaCep commented on public/app/features/alerting/unified/utils/routeTree.ts:
```diff
@@ -0,0 +1,119 @@
+/**
+ * Various helper functions to modify (immutably) the route tree, aka "notification policies"
+ *
+ * make sure to omit the "id" because Prometheus / Loki / Mimir will reject the payload
```
looks like this comment shouldn't be there

gillesdemey replied:
Yea good catch, I've added this comment in other places where this is more relevant 👍

VikaCep reviewed (commented):
Awesome work!!

gillesdemey committed:
small changes

soniaAguilarPeiron commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -0,0 +1,330 @@
+  }, [config?.route]);
+
+  // these are computed from the contactPoint and labels filter
+  const routesMatchingFilters = useMemo(() => {
```
💯

soniaAguilarPeiron commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -0,0 +1,330 @@
+  }
+
+  // edit, add, delete modals
+  const [addModal, openAddModal, closeAddModal] = useAddPolicyModal(receivers, handleAdd, updatingTree);
```
nice!❤️

soniaAguilarPeiron commented on public/app/features/alerting/unified/utils/notification-policies.ts:
```diff
@@ -0,0 +1,105 @@
+
+// Match does a depth-first left-to-right search through the route tree
+// and returns the matching routing nodes.
+function findMatchingRoutes<T extends Route>(root: T, labels: Label[]): T[] {
```
As we discussed yesterday, I think having a map with labels pointing to the list of the nodes in this tree would be a good improvement for this findMatching performance.

gillesdemey replied:
Yea this one bothers me a lot, it's incredibly slow for large trees and a large amount of alert groups. I'm considering dropping this feature for now and include it behind a feature flag so we have some time to optimize it without halting the release of this new policies view.

soniaAguilarPeiron reviewed (commented):
Amazing work! I think this is a big improvement for notifications policies! 🚀
I think we could add some follow-up notes somewhere:

- adding tests
- improving the performance of matching labels
- split code in different hooks / sub components so the code is more readable

gillesdemey committed:
simplify warnings

gillesdemey committed:
clean up policy component

gillesdemey committed:
use `createUrl`

gillesdemey commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -0,0 +1,330 @@
+    if (!rootRoute) {
+      return;
+    }
+    const newRouteTree = mergePartialAmRouteWithRouteTree(alertManagerSourceName ?? '', partialRoute, rootRoute);
```
No clue actually, it's retrieved from several places: query param and localStorage. I think it can be empty when there's no value in the localstorage and no query param.

gillesdemey commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -0,0 +1,330 @@
+        closeAddModal();
+        closeDeleteModal();
+      })
+      .finally(() => {
```
This is handled by https://github.com/grafana/grafana/blob/a12a0c73073c3509b54a40af42a56800235d3af6/public/app/features/alerting/unified/state/actions.ts#L547 which will dispatch an error and show a toast.

gillesdemey committed:
PR feedback

gillesdemey committed:
Merge branch 'main' into alerting/notification-policies-next

gillesdemey committed:
disable matching instances

This is just for now, will re-enable this with a feature flag later once we figure out how to do this properly with web workers

gillesdemey commented:
I've disabled the `x instances` shown on each policy for now since it crashes the page when the notification policy tree is large or when there are many firing instances.

The next steps for re-enabling these (in a separate PR) are:

*  Figure out how to use feature flags in the FE (I don't think these need to be defined in the BE)
* Use web workers to compute the matching instances so we don't block the main thread
* Figure out how to scale this up to more than one worker eventually (1 worker per policy is not feasible)

gillesdemey committed:
skip test for matching instances

gillesdemey committed:
feat: use label matchers for search

soniaAguilarPeiron reviewed (approved):
Awesome!🚀
I think we could add some follow-up notes somewhere:

- adding skipped tests (maybe adding to the todo list?)
- improving the performance of matching labels
- split code in different hooks / sub components so the code is more readable
- badge for errors in notification policy is red, and on contact points is yellow (that was decided by UX). We should be consistent and decide which one is the correct color.

soniaAguilarPeiron commented on packages/grafana-data/src/types/icon.ts:
```diff
@@ -4,6 +4,11 @@ export const availableIconsIndex = {
   github: true,
   gitlab: true,
   okta: true,
+  discord: true,
```
why do we need this in this PR?

gillesdemey replied:
You can load any icon from [unicons](https://iconscout.com/unicons/explore/line) but adding them here adds code completion to vscode.

gillesdemey replied:
Oh and because Discord is a supported contact point and we show an icon for it when we can :)

soniaAguilarPeiron reviewed (commented):
I think that is not clear how to create a new policy. The option is under the Edit button in the default policy. I think lots of users could get lost finding the button to create a new policy. Maybe we could add a button next to the edit button?
@dhalachliyski wdyt?

<img width="1466" alt="Screenshot 2023-02-24 at 10 23 57" src="https://user-images.githubusercontent.com/33540275/221141686-b7a4e230-b94c-4d60-b8ae-dda033a87633.png">

soniaAguilarPeiron reviewed (changes requested):
Almost ready to be approved 🙌
I added a comment regarding how to create a new notification policy. I think it's not clear where is the option for creating a new policy.

dhalachliyski commented:
I agree with @soniaAguilarPeiron's comment regarding the buttons. The primary action we expect should be to create a new policy. So let's make this the default visible button.
I think Editing and Deleting are closer in meaning, since they both modify the existing object you are interacting with. From this perspective, I think it is better to have these two button in the menu. Also, I think it would be better to visually separate the action button and the one showing the menu, because the menu actions are not modifications of the other one (for example if we had Save and Save As, it would have been OK if it was a Save button with an attached menu button).
Also, regarding "Add additional policy" and "Create nested policy" - I think from the user's perspective these two actions have absolutely the same end result - a new policy is created and this new policy appears on one level below the policy from which I triggered it. So let's use the same terminology for both. I'd advise we use the term "New" instead of "Add" (which suggests adding something that already exists) or "Create" (we've had the Create vs New debate over Slack recently). I'm note sure if "nested" or "child" is most descriptive, but for example we could name both actions "New nested policy".
I'm attaching a quick mockup for clarity.
<img width="1974" alt="image" src="https://user-images.githubusercontent.com/119334180/221166513-23607f60-faea-41d1-951b-ef70a5c03666.png">

VikaCep commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -127,13 +126,13 @@ const Policy: FC<PolicyComponentProps> = ({
-    () => findMatchingAlertGroups(routeTree, currentRoute, alertGroups),
-    [alertGroups, currentRoute, routeTree]
-  );
+  // const matchingAlertGroups = useMemo(
```
Can we add a comment or link https://github.com/grafana/grafana/pull/61952#issuecomment-1437010087 so someone that hasn't seen this PR is aware of why it's commented in the code? 
Another thought, would it be worth it to execute this only when the tree is small/there are not a lot of firing instances? Not sure how easy that would be to calculate though.

gillesdemey replied:
I've merged with main and hooked it up to the new feature toggles so we can ship the feature but behind a flag 👍

gillesdemey commented:
> badge for errors in notification policy is red, and on contact points is yellow (that was decided by UX). We should be consistent and decide which one is the correct color.

I'd love to figure this out, it always struck me as odd that it's yellow on the contact points page (it's definitely an error and needs to be dealt with, it's not working until then :))

gillesdemey committed:
PR feedback

gillesdemey committed:
Merge branch 'main' into alerting/notification-policies-next

gillesdemey committed:
Merge branch 'main' into alerting/notification-policies-next

gillesdemey committed:
hook up  matching instances to feature toggles

gillesdemey committed:
update tests

soniaAguilarPeiron reviewed (approved):
Great job! 🎉🚀

mckn reviewed (approved):
I'm not an expert on the alerting parts but as far as I can see it looks good!

konrad147 reviewed (approved):
Awesome job! 🎉  LGTM!

gillesdemey committed:
fix typo
